### ISSUE TITLE
Off-by-one error in array indexing

### ISSUE THREAD

**Issue by schlatterbeck** [robhagemans/pcbasic#182]
As suggested in private email I'm trying to debug a basic program using the python debugger.
When I use an array produced with DIM, I'm getting an off-by-one error when displaying the array using self._memory.arrays.to_list
Consider the following test program:
```
1 DIM J2(5,2)
2 J2(1,1)=-1
3 J2(1,2)=-1
4 PRINT J2(1,1);J2(1,2)
```

When I run this, it correctly prints "-1 -1". But in the debugger I get:
```
python3 -m pdb =pcbasic t.bas --interface=none -d
(Pdb) break /usr/lib/python3/dist-packages/pcbasic/basic/interpreter.py:105, struct.unpack_from('<H', token, 2)[0] == 4
Breakpoint 1 at /usr/lib/python3/dist-packages/pcbasic/basic/interpreter.py:105
(Pdb) c
> /usr/lib/python3/dist-packages/pcbasic/basic/interpreter.py(105)parse()
-> if self.tron:
(Pdb) p self._memory.arrays.to_list (b'J2!')
[[0.0, 0.0], [0.0, -1.0], [0.0, 0.0], [0.0, 0.0], [0.0, 0.0]]
(Pdb) c
-1 -1 
Ok 
```
This sets a conditional breakpoint in the interpreter loop and makes it stop in the Basic program above in line 4. When I display the array using arrays.to_list the python list is off-by one: The values are in the position where the Basic 1-based index would expect them. But I cannot display the value at J2(1,2) in python.
I don't know where the to_list method is used so I don't know if this bug has further problems down the road.

Note that the program I'm debugging uses fairly overDIMensioned arrays, so it's not an immediate problem (as long as I remember to use 1-based indexing when debugging in python)

**Comment by Marrin** [comment]
I'm not sure what you see as an off by one error here?  PC-BASIC arrays are 0 based by default unless you use `OPTION BASE 1`.

**Comment by schlatterbeck** [comment]
On Tue, Jun 21, 2022 at 03:12:37AM -0700, Marc Rintsch wrote:
> I'm not sure what you see as an off by one error here?  PC-BASIC
> arrays are 0 based by default unless you use `OPTION BASE 1`.

Interesting, when I was doing something in Basic everything was 1-based.
Why does my small example code work then? It does access the
array J2 produced with DIM at a coordinate that should not exist?

DIM J2(5,2)
...
J2(1,2)=-1

The index 2 in the second position should not be allowed if zero-based, then?

Note that I'm running an existing antenna-simulation program called mininec
https://github.com/Kees-PA3KJ/MiniNec

which uses 1-based indexing without any option and works fine.
I just cannot examine the last dimension of any array with
self._memory.arrays.to_list

If the DIM above uses 0-based indexing which would make the array of
size 6 X 3 then the self._memory.arrays.to_list at least does not list
the last index of any dimension as can be seen from the example I've given.

Ralf
-- 
Dr. Ralf Schlatterbeck                  Tel:   +43/2243/26465-16
Open Source Consulting                  www:   www.runtux.com
Reichergasse 131, A-3411 Weidling       email: ***@***.***

**Comment by Marrin** [comment]
`DIM A(2)` results in an array with 3 elements, with indices 0, 1, and 2.

From the documentation of `DIM`:
> `limit_0`, `limit_1`, ... are numeric expressions that specify the greatest index allowed at that position.

This doesn't change whether `OPTION BASE 0` or `OPTION BASE 1` was executed.

**Comment by schlatterbeck** [comment]
I see. 
self._memory.arrays.to_list
does not list the last dimension (the first index 6 and/or the second index 2) then.
This means when using the python debugger to debug a Basic program I cannot inspect the last dimension of any array.
I would expect
```
(Pdb) p self._memory.arrays.to_list (b'J2!')
[[0.0, 0.0, 0.0], [0.0, -1.0, -1.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]]
```
Compare this to the output that is produced, the third dimension in the inner array and the 6th dimension in the outer array are missing.

**Comment by Marrin** [comment]
I'm using the last version published on the Python Package Index and get a completely empty list in the debugger.  It looks a bit cumbersome anyway to set that breakpoint.

I would just extend the BASIC program with a subroutine writing the array contents to screen or into a file.  Or define a custom BASIC command.

**Comment by Marrin** [comment]
It looks like the two `range()` calls in the following code should go one dimension further:

https://github.com/robhagemans/pcbasic/blob/0e0a758d7dca91daf6bfe8c6111218f46967a9ae/pcbasic/basic/memory/arrays.py#L305

Calling `get()` gives the correct values:

```
(Pdb) p self._memory.arrays.get(b'J2!', [1,1])
b'!'[b'00008081' -1.0]
(Pdb) p self._memory.arrays.get(b'J2!', [1,2])
b'!'[b'00008081' -1.0]
```

**Comment by robhagemans** [comment]
Yes, looks like a bug. Thanks!

**Comment by Marrin** [comment]
@schlatterbeck GW-BASIC/PC-BASIC has a [RENUM](https://robhagemans.github.io/pcbasic/doc/2.0/#RENUM)ber command.

I guess I would port it to QBasic first to get rid of the unnecessary line numbers and replace the remaining ones with names and then starting to replace GOTOs with IF/ENDIF, and the different loop constructs and then the GOSUBs by SUBs and FUNCTIONs before porting it to some modern language.

**Comment by Marrin** [comment]
Sorry, I can't see any beauty in it and I think it is much more error prone to *not* convert it into some structured code first without any GOTOs.  You'll need to do that in the end anyway.

QBasic is Microsofts next step after GW-BASIC and it is almost 100% backwards compatible, i.e. can run GW-BASIC programs unchanged.  After removing unnecessary line numbers that are not a target of a GOTO, GOSUB, etc. one can eliminate GOTOs with multi line IF/ELIF/ELSE/ENDIF and loops.  With proper indentation.  The remaining line numbers than are subroutines/functions and can be put into SUB/FUNCTIONs.  Those conversions should be straight forward, there's very little room for errors that change the meaning of the code.

What about extending BASIC with a `_DUMP` command?

```python
#!/usr/bin/env python3
from pprint import pprint

import pcbasic


SOURCE = """\
1 DIM J2(5,2)
2 J2(1,1)=-1
3 J2(1,2)=-1:_DUMP("J2!()")
4 PRINT J2(1,1);J2(1,2)
"""


class Session(pcbasic.Session):
    def __init__(self, *args, **kwargs):
        pcbasic.Session.__init__(self, *args, extension=self, **kwargs)

    def dump(self, varname):
        print(varname, ":")
        pprint(self.get_variable(varname))


def main():
    with Session() as session:
        session.execute(SOURCE)
        session.execute("RUN")


if __name__ == "__main__":
    main()
```

Output (still with the bug in `_to_list()` of course):
```
b'J2!()' :
[[0.0, 0.0], [0.0, -1.0], [0.0, 0.0], [0.0, 0.0], [0.0, 0.0]]
-1 -1
```

One may even start to replace parts of the BASIC source code with functions to replace it, and this way moving the program piece by piece into Python.

**Comment by schlatterbeck** [comment]
On Tue, Jun 21, 2022 at 09:37:42AM -0700, Marc Rintsch wrote:
> Sorry, I can't see any beauty in it and I think it is much more error
> prone to *not* convert it into some structured code first without any
> GOTOs.  You'll need to do that in the end anyway.

Hmm, I'm doing just that. But I'm not going to another Basic dialect but
to Python. And I don't want to do an intermediate step to yet another
language. I need the debugger stuff just to search for bugs in *my* code
(comparing the control flow of the original code to my code).
I'm also trying to move all vector computations with multiple variables
to a single numpy array. And all variable duos that represent complex
numbers to complex number arithmetics.

Consider that (a complex 3d point):
707 X1=0
708 Y1=0
709 Z1=0
710 X2=0
711 Y2=0
712 Z2=0
...
729 S1=COS(S2)
730 S2=SIN(S2)
731 B1=F3*(S1*CR(I)-S2*CI(I))
732 B2=F3*(S1*CI(I)+S2*CR(I))
...
734 X1=X1+K*B1*CA(L)
735 X2=X2+K*B2*CA(L)
736 Y1=Y1+K*B1*CB(L)
737 Y2=Y2+K*B2*CB(L)
738 Z1=Z1+B1*CG(L)
739 Z2=Z2+B2*CG(L)

and that:
vec = np.zeros (3, dtype = complex)
...
kvec2 = np.array ([k, k, 1])
s = np.cos (s2) + 1j * np.sin (s2)
b = f3 * self.current * s
vec += kvec2 * b * wire.dirs

And you immediately see that it's sometimes necessary to verify that
code?

And we might want to move the discussion to direct mail: ***@***.***

Ralf
-- 
Dr. Ralf Schlatterbeck                  Tel:   +43/2243/26465-16
Open Source Consulting                  www:   www.runtux.com
Reichergasse 131, A-3411 Weidling       email: ***@***.***

**Comment by robhagemans** [comment]
Fixed with commit 525f9115
